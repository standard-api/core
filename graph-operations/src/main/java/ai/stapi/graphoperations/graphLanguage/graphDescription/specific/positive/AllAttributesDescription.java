package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive;

import ai.stapi.graphoperations.graphLanguage.GraphBaseTypes;
import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescriptionParameters;
import java.util.List;

public class AllAttributesDescription extends AbstractPositiveGraphDescription {

  public static final String SERIALIZATION_TYPE = "AllAttributesDescription";

  private AllAttributesDescriptionParameters allAttributesDescriptionParameters;

  protected AllAttributesDescription() {
  }

  public AllAttributesDescription(List<GraphDescription> childDescriptions) {
    super(AllAttributesDescription.SERIALIZATION_TYPE, GraphBaseTypes.VALUE_TYPE,
        childDescriptions);
  }

  public AllAttributesDescription(GraphDescription... childDescriptions) {
    super(AllAttributesDescription.SERIALIZATION_TYPE, GraphBaseTypes.VALUE_TYPE,
        childDescriptions);
  }

  @Override
  public GraphDescriptionParameters getParameters() {
    return this.allAttributesDescriptionParameters;
  }
}
