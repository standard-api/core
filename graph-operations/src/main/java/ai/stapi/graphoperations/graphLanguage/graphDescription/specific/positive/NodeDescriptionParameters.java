package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescriptionParameters;

public class NodeDescriptionParameters implements GraphDescriptionParameters {

  private String nodeType;

  private NodeDescriptionParameters() {

  }

  public NodeDescriptionParameters(String nodeType) {
    this.nodeType = nodeType;
  }

  public String getNodeType() {
    return nodeType;
  }
}