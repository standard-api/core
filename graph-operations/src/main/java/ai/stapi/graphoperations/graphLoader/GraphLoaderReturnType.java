package ai.stapi.graphoperations.graphLoader;

public enum GraphLoaderReturnType {
  GRAPH,
  OBJECT,
  SORT_OPTION,
  FILTER_OPTION
}
