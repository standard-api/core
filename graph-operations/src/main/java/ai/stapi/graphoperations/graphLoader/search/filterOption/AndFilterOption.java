package ai.stapi.graphoperations.graphLoader.search.filterOption;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class AndFilterOption extends AbstractCompositeFilterOption {

  public static final String STRATEGY = "and";

  private AndFilterOption() {
  }

  public AndFilterOption(FilterOption<?>... childFilterOptions) {
    this(Arrays.stream(childFilterOptions).collect(Collectors.toList()));
  }

  public AndFilterOption(List<FilterOption<?>> childFilterOptions) {
    super(AndFilterOption.STRATEGY, childFilterOptions);
  }
}
