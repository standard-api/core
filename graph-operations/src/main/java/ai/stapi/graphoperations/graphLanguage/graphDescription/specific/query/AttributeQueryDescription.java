package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.query;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.AbstractAttributeDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.AttributeDescriptionParameters;
import java.util.ArrayList;
import java.util.List;

public class AttributeQueryDescription extends AbstractAttributeDescription implements QueryGraphDescription {

  public static final String SERIALIZATION_TYPE = "AttributeQueryDescription";
  public static final String DESCRIBED_STRUCTURE_TYPE = "Irrelevant";

  private AttributeQueryDescription() {
  }

  public AttributeQueryDescription(
      AttributeDescriptionParameters parameters,
      GraphDescription... childDescriptions
  ) {
    super(SERIALIZATION_TYPE, DESCRIBED_STRUCTURE_TYPE, parameters, childDescriptions);
  }

  public AttributeQueryDescription(
      AttributeDescriptionParameters parameters,
      List<GraphDescription> childDescriptions
  ) {
    super(SERIALIZATION_TYPE, DESCRIBED_STRUCTURE_TYPE, parameters, childDescriptions);
  }

  public AttributeQueryDescription(
      String attributeName,
      List<GraphDescription> childDescriptions
  ) {
    super(SERIALIZATION_TYPE, DESCRIBED_STRUCTURE_TYPE, attributeName, childDescriptions);
  }

  public AttributeQueryDescription(
      String attributeName
  ) {
    super(SERIALIZATION_TYPE, DESCRIBED_STRUCTURE_TYPE, attributeName, new ArrayList<>());
  }
}
