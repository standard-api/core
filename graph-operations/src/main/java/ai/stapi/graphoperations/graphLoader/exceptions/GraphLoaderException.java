package ai.stapi.graphoperations.graphLoader.exceptions;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.AbstractEdgeDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.EdgeDescriptionParameters;
import ai.stapi.graphoperations.objectGraphLanguage.ObjectGraphMapping;
import ai.stapi.utils.Stringifier;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import ai.stapi.identity.UniqueIdentifier;
import java.util.List;

public class GraphLoaderException extends RuntimeException {

  public GraphLoaderException(String becauseMessage) {
    super("Cannot load graph, " + becauseMessage);
  }

  private GraphLoaderException(String becauseMessage, Throwable cause) {
    super("Cannot load graph, " + becauseMessage, cause);
  }

  public static GraphLoaderException becauseThereIsNoGraphElementWithProvidedUuid(
      UniqueIdentifier uuid,
      GraphDescription graphDescription
  ) {
    return new GraphLoaderException(
        "because specified graph element does not exist in graph." +
            "\nUUID: " + uuid +
            "\nGraphDescription: " + graphDescription.getClass().getSimpleName() +
            "\n" + Stringifier.convertToString(graphDescription)
    );
  }

  public static GraphLoaderException becauseThereIsMoreThanOneNodeDescriptionsOnEdgeDescription(
      UniqueIdentifier id, String type) {
    return new GraphLoaderException(
        "because there is more than one node descriptions on edge description." +
            "\nUnique Identifier: " + id +
            "\nEdge type: " + type
    );
  }

  public static GraphLoaderException becauseEdgeDescriptionHadMoreChildNodeDescriptions(
      AbstractEdgeDescription edgeDescription,
      List<String> nodeTypes
  ) {
    var param = (EdgeDescriptionParameters) edgeDescription.getParameters();
    return new GraphLoaderException(
        "because edge description had more child node descriptions." +
            "\nEdge Description ClassName: " + edgeDescription.getClass().getSimpleName() +
            "\nEdge type: " + param.getEdgeType() +
            "\nNode Types: " + String.join(", ", nodeTypes)
    );
  }

  public static GraphLoaderException becauseProvidedOGMDidNotStartWithObjectOGM(
      ObjectGraphMapping objectGraphMapping) {
    return new GraphLoaderException(
        "because provided OGM did not start with object OGM." +
            "\nOGM Class: " + objectGraphMapping.getClass().getSimpleName()
    );
  }
}
