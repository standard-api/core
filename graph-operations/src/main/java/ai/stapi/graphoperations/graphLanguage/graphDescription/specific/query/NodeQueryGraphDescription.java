package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.query;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.AbstractNodeDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.NodeDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.NodeDescriptionParameters;
import ai.stapi.graphoperations.graphLoader.search.SearchQueryParameters;
import java.util.List;

public class NodeQueryGraphDescription extends AbstractNodeDescription
    implements GraphElementQueryDescription {

  public static final String SERIALIZATION_TYPE = "72ba0bef-5f38-489b-bbbf-8212191f1f92";

  private SearchQueryParameters searchQueryParameters;

  protected NodeQueryGraphDescription() {
    super();
  }

  public NodeQueryGraphDescription(NodeDescription nodeDescription) {
    super(
        SERIALIZATION_TYPE,
        (NodeDescriptionParameters) nodeDescription.getParameters(),
        nodeDescription.getChildGraphDescriptions()
    );
    this.searchQueryParameters = new SearchQueryParameters();
  }

  public NodeQueryGraphDescription(
      NodeDescriptionParameters parameters,
      SearchQueryParameters searchQueryParameters,
      List<GraphDescription> childDeclarations
  ) {
    super(SERIALIZATION_TYPE, parameters, childDeclarations);
    this.parameters = parameters;
    this.searchQueryParameters = searchQueryParameters;
  }

  public NodeQueryGraphDescription(
      NodeDescriptionParameters parameters,
      GraphDescription... childDeclarations
  ) {
    super(SERIALIZATION_TYPE, parameters, childDeclarations);
    this.parameters = parameters;
    this.searchQueryParameters = new SearchQueryParameters();
  }

  public NodeQueryGraphDescription(
      NodeDescriptionParameters parameters
  ) {
    super(SERIALIZATION_TYPE, parameters);
    this.searchQueryParameters = new SearchQueryParameters();
  }

  public NodeQueryGraphDescription(
      NodeDescriptionParameters parameters,
      SearchQueryParameters searchQueryParameters,
      GraphDescription... childDeclarations
  ) {
    super(SERIALIZATION_TYPE, parameters, childDeclarations);
    this.searchQueryParameters = searchQueryParameters;
  }

  @Override
  public SearchQueryParameters getSearchQueryParameters() {
    return this.searchQueryParameters;
  }
}
