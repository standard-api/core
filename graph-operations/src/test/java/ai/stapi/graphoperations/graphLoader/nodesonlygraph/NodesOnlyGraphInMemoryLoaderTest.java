package ai.stapi.graphoperations.graphLoader.nodesonlygraph;

import ai.stapi.graphoperations.fixtures.AttributeTypes;
import ai.stapi.graphoperations.fixtures.testsystem.TestSystemModelDefinitionsLoader;
import ai.stapi.identity.UniversallyUniqueIdentifier;
import ai.stapi.graph.repositorypruner.RepositoryPruner;
import ai.stapi.graphoperations.fixtures.NodesOnlyGraphFixturesProvider;
import ai.stapi.graphoperations.fixtures.model.GraphLoaderTestDefinitionsLoader;
import ai.stapi.graphoperations.synchronization.GraphSynchronizer;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.AllAttributesDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.NodeDescriptionParameters;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.query.AttributeQueryDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.query.NodeQueryGraphDescription;
import ai.stapi.graphoperations.graphLoader.GraphLoaderReturnType;
import ai.stapi.graphoperations.graphLoader.exceptions.GraphLoaderException;
import ai.stapi.graphoperations.graphLoader.inmemory.InMemoryGraphLoader;
import ai.stapi.graphoperations.graphLoader.search.SearchQueryParameters;
import ai.stapi.graphoperations.graphLoader.search.filterOption.AndFilterOption;
import ai.stapi.graphoperations.graphLoader.search.filterOption.ContainsFilterOption;
import ai.stapi.graphoperations.graphLoader.search.filterOption.EndsWithFilterOption;
import ai.stapi.graphoperations.graphLoader.search.filterOption.EqualsFilterOption;
import ai.stapi.graphoperations.graphLoader.search.filterOption.GreaterThanFilterOption;
import ai.stapi.graphoperations.graphLoader.search.filterOption.GreaterThanOrEqualFilterOption;
import ai.stapi.graphoperations.graphLoader.search.filterOption.LowerThanFilterOption;
import ai.stapi.graphoperations.graphLoader.search.filterOption.LowerThanOrEqualsFilterOption;
import ai.stapi.graphoperations.graphLoader.search.filterOption.NotFilterOption;
import ai.stapi.graphoperations.graphLoader.search.filterOption.OrFilterOption;
import ai.stapi.graphoperations.graphLoader.search.filterOption.StartsWithFilterOption;
import ai.stapi.graphoperations.graphLoader.search.paginationOption.OffsetPaginationOption;
import ai.stapi.graphoperations.graphLoader.search.sortOption.AscendingSortOption;
import ai.stapi.graphoperations.graphLoader.search.sortOption.DescendingSortOption;
import ai.stapi.test.schemaintegration.SchemaIntegrationTestCase;
import ai.stapi.test.schemaintegration.StructureDefinitionScope;
import java.util.List;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;
import org.springframework.beans.factory.annotation.Autowired;

@StructureDefinitionScope({
    GraphLoaderTestDefinitionsLoader.SCOPE,
    TestSystemModelDefinitionsLoader.SCOPE
})
class NodesOnlyGraphInMemoryLoaderTest extends SchemaIntegrationTestCase {

  @Autowired
  private InMemoryGraphLoader inMemoryGraphLoader;

  @BeforeAll
  public static void beforeAll(
      @Autowired GraphSynchronizer graphSynchronizer,
      @Autowired NodesOnlyGraphFixturesProvider nodesOnlyGraphFixturesProvider,
      @Autowired RepositoryPruner repositoryPruner
  ) {
    repositoryPruner.prune();
    graphSynchronizer.synchronize(nodesOnlyGraphFixturesProvider.getFixturesGraph());
  }

  @Test
  void itShouldFindAllWhenQuerySearchParamsAreEmpty() {
    var searchParam = SearchQueryParameters.from();
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(
                NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenUnsortedGraphElementsApproved(actual);
  }

  @Test
  void itShouldSortDescendingByQuantity() {
    var descendingSortOption = new DescendingSortOption(AttributeTypes.EXAMPLE_QUANTITY);
    var searchParam = SearchQueryParameters.from(descendingSortOption);
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenUnsortedGraphElementsApproved(actual);
  }

  @Test
  void itShouldSortByStringAttributeAscAndThanQuantityDesc() {
    var ascendingStringAttributeSortOption = new AscendingSortOption(AttributeTypes.EXAMPLE_STRING_ATTRIBUTE);
    var descendingNameSortOption = new DescendingSortOption(AttributeTypes.EXAMPLE_QUANTITY);
    var searchParam =
        SearchQueryParameters.from(ascendingStringAttributeSortOption, descendingNameSortOption);
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenUnsortedGraphElementsApproved(actual);
  }

  @Test
  void itShouldIgnoreSortOptionWithNonexistingAttributeName() {
    var nonexistingSortOption = new DescendingSortOption("nonexising_attribute_name");
    var searchParam = SearchQueryParameters.from(nonexistingSortOption);
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(
                NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenUnsortedGraphElementsApproved(actual);
  }

  @Test
  void itShouldIgnoreSortOptionWithNonexistingAttributeNameAndThanByQuantityAsc() {
    var nonexistingSortOption = new DescendingSortOption("nonexising_attribute_name");
    var ascQuantitySortOption = new AscendingSortOption(AttributeTypes.EXAMPLE_QUANTITY);
    var searchParam = SearchQueryParameters.from(nonexistingSortOption, ascQuantitySortOption);
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenUnsortedGraphElementsApproved(actual);
  }

  @Test
  void itShouldReturnEmptyListWhenFilterOptionWithNonexistingAttributeName() {
    var nonexistingFilterOption = new EqualsFilterOption<>(
        "nonexising_attribute_name",
        "irrelevant value"
    );
    var searchParam = SearchQueryParameters.from(nonexistingFilterOption);
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AllAttributesDescription()
        )
    );
    Assertions.assertEquals(0, actual.size());
  }

  @Test
  void itShouldFilterWithEqualsFilterOption() {
    var equalsFilterOption = new EqualsFilterOption<>(
        AttributeTypes.EXAMPLE_STRING_ATTRIBUTE,
        "Attribute Value"
    );
    var searchParam = SearchQueryParameters.from(equalsFilterOption);
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(
                NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenUnsortedGraphElementsApproved(actual);
  }

  @Test
  void itShouldFilterWithGreaterThanFilterOption() {
    var equalsFilterOption = new GreaterThanFilterOption<>(
        AttributeTypes.EXAMPLE_QUANTITY,
        653
    );
    var searchParam = SearchQueryParameters.from(equalsFilterOption);
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenUnsortedGraphElementsApproved(actual);
  }

  @Test
  void itShouldFilterWithGreaterThanOrEqualsFilterOption() {
    var equalsFilterOption = new GreaterThanOrEqualFilterOption<>(
        AttributeTypes.EXAMPLE_QUANTITY,
        653
    );
    var searchParam = SearchQueryParameters.from(equalsFilterOption);
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(
                NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenUnsortedGraphElementsApproved(actual);
  }

  @Test
  void itShouldFilterWithLowerThanFilterOption() {
    var equalsFilterOption = new LowerThanFilterOption<>(
        AttributeTypes.EXAMPLE_QUANTITY,
        653
    );
    var searchParam = SearchQueryParameters.from(equalsFilterOption);
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(
                NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenUnsortedGraphElementsApproved(actual);
  }

  @Test
  void itShouldFilterWithLowerThanOrEqualFilterOption() {
    var equalsFilterOption = new LowerThanOrEqualsFilterOption<>(
        AttributeTypes.EXAMPLE_QUANTITY,
        653
    );
    var searchParam = SearchQueryParameters.from(equalsFilterOption);
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(
                NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenUnsortedGraphElementsApproved(actual);
  }

  @Test
  void itShouldFilterWithEqualsFilterOptionAndSortByQuantityAsc() {
    var equalsFilterOption = new EqualsFilterOption<>(
        AttributeTypes.EXAMPLE_STRING_ATTRIBUTE,
        "Attribute Value"
    );
    var quantitySortOption = new AscendingSortOption(AttributeTypes.EXAMPLE_QUANTITY);
    var searchParam = SearchQueryParameters.from(equalsFilterOption, quantitySortOption);
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(
                NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenUnsortedGraphElementsApproved(actual);
  }

  @Test
  void itShouldFilterByTwoFiltersAndSortByQuantityAsc() {
    var equalsFilterOption = new EqualsFilterOption<>(
        AttributeTypes.EXAMPLE_STRING_ATTRIBUTE,
        "Attribute Value"
    );
    var greaterFilterOption = new GreaterThanFilterOption<>(
        AttributeTypes.EXAMPLE_QUANTITY,
        20
    );
    var quantitySortOption = new AscendingSortOption(AttributeTypes.EXAMPLE_QUANTITY);
    var searchParam = SearchQueryParameters.from(
        equalsFilterOption, greaterFilterOption, quantitySortOption
    );
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(
                NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenUnsortedGraphElementsApproved(actual);
  }

  @Test
  void itShouldFilterByAndFilterOptions() {
    var equalsFilterOption = new EqualsFilterOption<>(
        AttributeTypes.EXAMPLE_STRING_ATTRIBUTE,
        "Attribute Value"
    );
    var greaterFilterOption = new GreaterThanFilterOption<>(
        AttributeTypes.EXAMPLE_QUANTITY,
        20
    );
    var andFilterOption = new AndFilterOption(equalsFilterOption, greaterFilterOption);
    var searchParam = SearchQueryParameters.from(andFilterOption);
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(
                NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenUnsortedGraphElementsApproved(actual);
  }

  @Test
  void itShouldFilterByOrFilterOptions() {
    var equalsFilterOption1 = new EqualsFilterOption<>(
        AttributeTypes.EXAMPLE_STRING_ATTRIBUTE,
        "Attribute Value"
    );
    var equalsFilterOption2 = new EqualsFilterOption<>(
        AttributeTypes.EXAMPLE_STRING_ATTRIBUTE,
        "Unique Attribute Value"
    );
    var orFilterOption = new OrFilterOption(List.of(equalsFilterOption1, equalsFilterOption2));
    var searchParam = SearchQueryParameters.from(orFilterOption);
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(
                NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenUnsortedGraphElementsApproved(actual);
  }

  @Test
  void itShouldFilterByNotFilterOptions() {
    var equalsFilterOption = new EqualsFilterOption<>(
        AttributeTypes.EXAMPLE_STRING_ATTRIBUTE,
        "Attribute Value"
    );
    var notFilterOption = new NotFilterOption(equalsFilterOption);
    var searchParam = SearchQueryParameters.from(notFilterOption);
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(
                NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenUnsortedGraphElementsApproved(actual);
  }

  @Test
  void itShouldFilterByLikeFilterOptions_WhenSearchedStringIsInMiddle() {
    var likeFilterOption = new ContainsFilterOption(
        AttributeTypes.EXAMPLE_STRING_ATTRIBUTE,
        "Other"
    );
    var searchParam = SearchQueryParameters.from(likeFilterOption);
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenUnsortedGraphElementsApproved(actual);
  }

  @Test
  void itShouldFilterByLikeFilterOptions_WhenSearchedStringIsAtStart() {
    var likeFilterOption = new ContainsFilterOption(
        AttributeTypes.EXAMPLE_STRING_ATTRIBUTE,
        "Attribute"
    );
    var searchParam = SearchQueryParameters.from(likeFilterOption);
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(
                NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenUnsortedGraphElementsApproved(actual);
  }

  @Test
  void itShouldFilterByLikeFilterOptions_WhenSearchedStringIsAtEnd() {
    var likeFilterOption = new ContainsFilterOption(
        AttributeTypes.EXAMPLE_NAME,
        "Name 4"
    );
    var searchParam = SearchQueryParameters.from(likeFilterOption);
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(
                NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenUnsortedGraphElementsApproved(actual);
  }

  @Test
  void itShouldFilterByStartsWithFilterOption() {
    var startsWithFilterOption = new StartsWithFilterOption(
        AttributeTypes.EXAMPLE_STRING_ATTRIBUTE,
        "Attribute"
    );
    var searchParam = SearchQueryParameters.from(startsWithFilterOption);
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenUnsortedGraphElementsApproved(actual);
  }

  @Test
  void itShouldFilterByEndsWithFilterOption() {
    var endsWithFilterOption = new EndsWithFilterOption(
        AttributeTypes.EXAMPLE_NAME,
        "Name 4"
    );
    var searchParam = SearchQueryParameters.from(endsWithFilterOption);
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(
                NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenUnsortedGraphElementsApproved(actual);
  }

  @Test
  void itShouldPaginateResult() {
    var paginationOption = new OffsetPaginationOption(2, 3);
    var ascendingSortOption = new AscendingSortOption(AttributeTypes.EXAMPLE_NAME);
    var searchParam = SearchQueryParameters.from(ascendingSortOption, paginationOption);
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(
                NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AllAttributesDescription()
        )
    );
    this.thenUnsortedGraphElementsApproved(actual);
  }

  @Test
  void itShouldPaginateResultAndMapToObject() {
    var paginationOption = new OffsetPaginationOption(2, 2);
    var ascendingSortOption = new AscendingSortOption(AttributeTypes.EXAMPLE_NAME);
    var searchParam = SearchQueryParameters.from(ascendingSortOption, paginationOption);
    var actual = this.inMemoryGraphLoader.find(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AttributeQueryDescription(AttributeTypes.EXAMPLE_NAME)
        ),
        Object.class,
        GraphLoaderReturnType.OBJECT
    );
    this.thenObjectApproved(actual.getData());
  }

  @Test
  void itShouldKeepOnlySpecifiedAttributes() {
    var actual = this.inMemoryGraphLoader.findAsTraversable(
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            SearchQueryParameters.from(),
            new AttributeQueryDescription(AttributeTypes.EXAMPLE_NAME)
        )
    );
    this.thenUnsortedGraphElementsApproved(actual);
  }

  @Test
  void itShouldThrowExceptionWhenGettingNodeAndItsNotPresent() {
    var searchParam = SearchQueryParameters.from();
    Executable throwable = () -> this.inMemoryGraphLoader.getAsTraversable(
        UniversallyUniqueIdentifier.fromString("576a533b-5764-401b-876c-9df37461bf90"),
        new NodeQueryGraphDescription(
            new NodeDescriptionParameters(NodesOnlyGraphFixturesProvider.EXAMPLE_ENTITY_NODE_TYPE),
            searchParam,
            new AllAttributesDescription()
        )
    );

    this.thenExceptionMessageApproved(GraphLoaderException.class, throwable);
  }
}
