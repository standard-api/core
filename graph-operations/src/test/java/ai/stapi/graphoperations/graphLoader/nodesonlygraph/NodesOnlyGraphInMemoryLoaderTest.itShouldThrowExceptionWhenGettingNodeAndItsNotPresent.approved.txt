Cannot load graph, because specified graph element does not exist in graph.
UUID: 576a533b-5764-401b-876c-9df37461bf90
GraphDescription: NodeQueryGraphDescription